File: app/src/main/java/com/ahmednts/scylla/MyApplication.java
package com.ahmednts.scylla;

import android.app.Application;
import android.os.StrictMode;
import com.crashlytics.android.Crashlytics;
import io.fabric.sdk.android.Fabric;
import timber.log.Timber;

/**
 * Created by AhmedNTS on 7/29/2017.
 */

public class MyApplication
    extends Application {

  private static MyApplication INSTANCE;

  public static MyApplication getInstance() {
    return INSTANCE;
  }

  public MyApplication() {
    super();
    INSTANCE = this;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    Fabric.with(this, new Crashlytics());

    //if (LeakCanary.isInAnalyzerProcess(this)) {
    //  // This process is dedicated to LeakCanary for heap analysis.
    //  // You should not init your app in this process.
    //  return;
    //}
    //
    //LeakCanary.install(this);

    Timber.plant(new Timber.DebugTree());

    setStrictMode();
  }

  private void setStrictMode() {
    if (BuildConfig.DEBUG) {
      StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder().detectDiskReads()
          .detectDiskWrites()
          .detectNetwork()   // or .detectAll() for all detectable problems
          .penaltyLog()
          .build());
      StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder().detectLeakedSqlLiteObjects()
          .detectLeakedClosableObjects()
          .penaltyLog()
          .penaltyDeath()
          .build());
    }
  }
}

File: app/src/main/java/com/ahmednts/scylla/MyApplication.kt
package com.ahmednts.scylla

import android.app.Application
import android.os.StrictMode
import com.ahmednts.scylla.utils.AppLogger
import io.reactivex.internal.functions.Functions
import io.reactivex.plugins.RxJavaPlugins

/**
 * Created by AhmedNTS on 7/29/2017.
 */

class MyApplication : Application() {
  companion object {
    private lateinit var INSTANCE: MyApplication

    @JvmStatic
    val instance
      get() = INSTANCE
  }

  init {
    INSTANCE = this
  }

  override fun onCreate() {
    super.onCreate()
    setStrictMode()

    AppLogger.setup()
    RxJavaPlugins.setErrorHandler(Functions.emptyConsumer())
  }

  private fun setStrictMode() {
    if (BuildConfig.DEBUG) {
      StrictMode.setThreadPolicy(
          StrictMode.ThreadPolicy.Builder().detectDiskReads().detectDiskWrites().detectNetwork()   // or .detectAll() for all detectable problems
              .penaltyLog().build()
      )
      StrictMode.setVmPolicy(
          StrictMode.VmPolicy.Builder().detectLeakedSqlLiteObjects().detectLeakedClosableObjects().penaltyLog().penaltyDeath().build()
      )
    }
  }
}
